drivers/pinctrl/pinctrl_wch_afio_20x_30x.c

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2024 MASSDRIVER EI (massdriver.space)
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/dt-bindings/pinctrl/ch32v20x_30x-pinctrl.h>
+
+#include <ch32fun.h>
+
+static GPIO_TypeDef *const wch_afio_pinctrl_regs[] = {
+	(GPIO_TypeDef *)DT_REG_ADDR(DT_NODELABEL(gpioa)),
+	(GPIO_TypeDef *)DT_REG_ADDR(DT_NODELABEL(gpiob)),
+	(GPIO_TypeDef *)DT_REG_ADDR(DT_NODELABEL(gpioc)),
+	(GPIO_TypeDef *)DT_REG_ADDR(DT_NODELABEL(gpiod)),
+#if DT_NODE_EXISTS(DT_NODELABEL(gpioe))
+	(GPIO_TypeDef *)DT_REG_ADDR(DT_NODELABEL(gpioe)),
+#endif
+};
+
+int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt, uintptr_t reg)
+{
+	int i;
+
+	for (i = 0; i < pin_cnt; i++, pins++) {
+		uint8_t port = (pins->config >> CH32V20X_V30X_PINCTRL_PORT_SHIFT) & 0x07;
+		uint8_t pin = (pins->config >> CH32V20X_V30X_PINCTRL_PIN_SHIFT) & 0x0F;
+		uint8_t bit0 = (pins->config >> CH32V20X_V30X_PINCTRL_RM_BASE_SHIFT) & 0x1F;
+		uint8_t pcfr_id = (pins->config >> CH32V20X_V30X_PINCTRL_PCFR_ID_SHIFT) & 0x01;
+		uint8_t remap = (pins->config >> CH32V20X_V30X_PINCTRL_RM_SHIFT) & 0x3;
```

Comment:
It is also possible to only define the mask (see other comment about `CH32V20X_V30X_PINCTRL_PORT_SHIFT` & co) and then use `FIELD_GET()` if desired:
https://github.com/zephyrproject-rtos/zephyr/blob/2dcb61858b2cb2294bedca3b556fe210213a63d5/include/zephyr/sys/util_macro.h#L111

Reply:
Will look into it :+1:

Reply:
Done.